segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/WriteOperation.java

```diff
@@ -105,46 +104,44 @@ public CompletableFuture<Void> call() {
-                                                                                                postCommit(), chunkedSegmentStorage.getExecutor())
-                                                                                        .exceptionally(this::handleException),
-                                                                        chunkedSegmentStorage.getExecutor())
-                                                                .thenRunAsync(this::logEnd, chunkedSegmentStorage.getExecutor()),
-                                                chunkedSegmentStorage.getExecutor());
-                            }, chunkedSegmentStorage.getExecutor());
-                }, chunkedSegmentStorage.getExecutor())
-                .exceptionally(ex -> (Void) handleException(ex));
+        return ChunkedSegmentStorage.tryWith(
+            chunkedSegmentStorage.getMetadataStore().beginTransaction(false, handle.getSegmentName()), txn -> {
+                didSegmentLayoutChange = false;
+
+                // Retrieve metadata.
+                return txn.get(streamSegmentName).thenComposeAsync(storageMetadata -> {
+                    segmentMetadata = (SegmentMetadata) storageMetadata;
+                    // Validate preconditions.
+                    checkState();
+
+                    isSystemSegment = chunkedSegmentStorage.isStorageSystemSegment(segmentMetadata);
+
+                    // Check if this is a first write after ownership changed.
+                    isFirstWriteAfterFailover = segmentMetadata.isOwnershipChanged();
+
+                    lastChunkMetadata.set(null);
+                    chunkHandle = null;
+                    bytesRemaining.set(length);
+                    currentOffset.set(offset);
+
+                    // Get the last chunk segmentMetadata for the segment.
+
+                    return getLastChunk(txn).thenComposeAsync(
+                        v -> writeData(txn).thenComposeAsync(
+                            vv -> commit(txn).thenApplyAsync(vvvv -> postCommit(), chunkedSegmentStorage.getExecutor())
+                                             .exceptionally(e -> {
+                                                 throw handleException(e);
+                                             }),
+                            chunkedSegmentStorage.getExecutor())
+                                           .thenRunAsync(this::logEnd, chunkedSegmentStorage.getExecutor()),
+                        chunkedSegmentStorage.getExecutor());
+                }, chunkedSegmentStorage.getExecutor());
+            }, chunkedSegmentStorage.getExecutor()).exceptionally(ex -> {
+                throw handleException(ex);
+            });
     }
 
-    private Object handleException(Throwable e) {
+    private RuntimeException handleException(Throwable e) {
         log.debug("{} write - exception op={}, segment={}, offset={}, length={}.",
```

Comment:
These changes seem to be unrelated with delayed future cancelation..

Reply:
Reverting this results in compile errors for me.